contract/src/airdrop/prepare.js

```diff
@@ -0,0 +1,267 @@
+import { M } from '@endo/patterns';
+import { makeDurableZone } from '@agoric/zone/durable.js';
+import { E } from '@endo/far';
+import { Far } from '@endo/marshal';
+import { AmountMath } from '@agoric/ertp';
+import { TimeMath, TimestampShape } from '@agoric/time';
+import '../../types.js'
+
+/**
+ * @typedef {object} EpochDetails
+ * @property {bigint} windowLength Length of epoch in seconds. This value is used by the contract's timerService to schedule a wake up that will fire once all of the seconds in an epoch have elapsed
+ * @property {bigint} earlyClaimBonus The total number of tokens recieved by each user who claims during a particular epoch.
+ * @property {bigint} index The index of a particular epoch.
+ * @property {number} inDays Length of epoch formatted in total number of days
+ */
+
+const startupAssertion = (arg, keyName) =>
+  assert(
+    arg,
+    `Contract has been started without required property: ${keyName}.`,
+  );
+
+const makeWaker = (name, func) => {
+  return Far(name, {
+    wake: timestamp => func(timestamp),
+  });
+};
+
+// const createWakeup = async (timer, wakeUpTime, timeWaker, cancelTokenMaker) => {
+//   const cancelToken = cancelTokenMaker();
+//   await E(timer).setWakeup(wakeUpTime, timeWaker, cancelToken);
+// };
+const makeCancelTokenMaker = name => {
+  let tokenCount = 1;
+
+  return () => Far(`cancelToken-${name}-${(tokenCount += 1)}`, {});
+};
+
+/**
+ *
+ * @typedef {object} ContractTerms
+ * @property {import('@endo/marshal').RemotableObject} AirdropUtils
+ * @property {bigint} startTime
+ * @property {bigint} basePayoutQuantity
+ * @property {{Keyword<Brand>}} brands
+ * @property {{Keyword<issuer>}} issuers
+ */
+
+/**
+ * @param {ZCF} zcf
+ * @param {{ purse: Purse, timer: import('@agoric/swingset-vat/tools/manual-timer.js').TimerService,}} privateArgs
+ * @param {import('@agoric/vat-data').Baggage} baggage
+ */
+export const start = async (zcf, privateArgs, baggage) => {
+  const zone = makeDurableZone(baggage, 'rootZone');
+
+  const terms = zcf.getTerms();
+
+  /** @type {ContractTerms} */
+  const {
+    AirdropUtils,
```

Comment:
what vat is `AirdropUtils` expected to come from? Why not make it part of the contract?

Reply:
When this contract is in production-ready form, the claim invitation will contain a function for verifying whether or not a particular user is eligible to claim. To pull this off, the contract's `start` (or `prepare`) function will be given the root hash of a merkle tree which has been constructed using each address that meets the eligibility criteria. 

In addition to constructing the root hash, a proof will be constructed for each address.

**Each of these operations is handled off-chain**.

At the time of claiming, users will pass in the proof associated with their address so that it can be verified against the root hash provided to the contract at start time.

The logic for verifying a user's proof against the contract's merkle root can come in a variety of forms, and so I thought it would be beneficial to give airdrop creators the ability to supply a proof verification method into the contract at start time, rather than having this be a static function predefined within the contract, and so this is where the idea to pass the contract a utility module was born.

I realize that implementation you've analyzed contains no indication of the flow I just discussed but I still felt it was important information to pass along. I also want to make it known that this utility module still needs to be thought about a bit more so that the contract is generic enough to customize, while making it difficult for airdrop creators to end up foot-gunning situation as a result of being handed this power.

### questions:

1. is it viable to pass a shared interface into a contract in such a way that the certain aspects of the contract logic can be customized at start time?
2. Should this hash be passed in as `customTerms` or `privateTerms`?

Reply:
> Should this hash be passed in as customTerms or privateTerms?

Seems like it's not sensitive, so `customTerms`.

Reply:
> is it viable to pass a shared interface into a contract in such a way that the certain aspects of the contract logic can be customized at start time?

Well, the object has to be in _some_ vat. We avoid making new objects in the bootstrap vat because it's not upgradeable and we aim to phase it out. So that leaves putting it in another contract. Is that the plan?

Another possibility is to pass in JS source in a string and use `Compartment.evaluate()`.

Reply:
parameterizing AirdropUtils is more complex than not. Is there an actual requirement? If not, [YouArentGonnaNeedIt](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).

Reply:
This was removed [here](https://github.com/Agoric/agoric-sdk/compare/master...tgrecojs:agoric-sdk:tg-durable-airdrop#diff-43a500ec9246338b6db7d30c6a36dea1c11f01fa6e7405678e942d2e8ff52c9c)